p2p/service.py

```diff
@@ -94,8 +97,21 @@ def __init__(self, token: CancelToken=None) -> None:
             if finished_callback is not None:
                 finished_callback(self)
 
+    def schedule_child_service(self, child_service: 'BaseService') -> None:
+        """
+        Schedule a child service to run with the main service. On shutdown the main service
+        will first wait until all child services cleaned up before running its own cleanup
+        """
+        self._child_services.append(child_service)
+
+    def _run_child_services(self) -> None:
+        for service in self._child_services:
+            asyncio.ensure_future(service.run())
+
     async def cleanup(self) -> None:
         """Run the service's _cleanup() coroutine and sets the cleaned_up event."""
+
+        await asyncio.gather(*[svc.cleaned_up.wait() for svc in self._child_services])
```

Comment:
@gsalgado do you think it is important that we wait on the child service cleanup first or could we do that in parallel as in:

```python
await asyncio.gather(*[svc.cleaned_up.wait() for svc in self._child_services], self._cleanup())
```

Reply:
I think in parallel is fine as long as it is clearly documented that we don't wait for them in any specific order